### DISCUSSION TITLE
Is it possible to have custom context menus (browser)?

### DISCUSSION THREAD

**Discussion by amarcolini** [nacular/doodle#50] [Q&A] [answered]
On the browser, all nodes have a context menu event and a default context menu provided by the browser. In my application I want to create a custom context menu for some Views that appears on right click. Unfortunately, due to Doodle being platform-independent, there is no support for removing the default browser context menu. Is this the case, or is there some other workaround?

**Comment by amarcolini** [comment]
Potential solutions I thought of:
 - Give each View a unique id visible from the DOM so that the HTMLElement of the View is accessible via `window.document.getElementById`
 - An extension property or function unique to the browser platform that gives access to the HTMLElement or to event listeners
 - An injectable Module that contains a map of each View to its corresponding HTMLElement or that can add event listeners to it

**Reply by amarcolini** [reply]
So as it turns out it is possible to use the `AccessibilityModule` and assign an id to `View.accessibilityLabel`, allowing HTMLElements to be accessed by `window.document.querySelector("div[aria-label='unique_view_id']")`. This is more of a temporary workaround, however, since it wouldn't be very good accessibility-wise.

**Reply by pusolito** [reply]
Have you tried registering a
`PointerListener` that consumes the event when [Button2](https://nacular.github.io/doodle-api/core/io.nacular.doodle.system/-system-pointer-event/-button/index.html) is the only value in `event.buttons`? Then you could use the [PopupManager](https://nacular.github.io/doodle/docs/modals) to show your own context menu that is relative to the pointer location (essentially `clickedView.toAbsolute(event.location)`).

**Reply by amarcolini** [reply]
I have tried consuming the released event, but it seems to not work. The PopupManager suggestion is great though - I hadn't thought of it for something like that. I think the reason for it not working is that contextmenu is its own event in the DOM.

**Reply by pusolito** [reply]
Have you tried consuming the pressed event and then triggering your menu on the released event?

**Reply by amarcolini** [reply]
It looks like that doesn't work either. The only solution I found to work was to consume the "contextmenu" event directly from the HTMLElements, which I got through aria-labels on empty child Labels. My solution looks like this:

```kotlin
import io.nacular.doodle.controls.text.Label
import io.nacular.doodle.core.View
import io.nacular.doodle.geometry.Size
import kotlinx.browser.window
import org.w3c.dom.Element

abstract class HTMLView : View() {
    companion object {
        private var count = 0
    }

    private val id = count++
    private val label = Label().apply {
        accessibilityLabel = "$id"
        size = Size.Empty
        visible = true
    }

    init {
        children += label
    }

    val htmlElement: Element? get() = window.document.querySelector("div[aria-label='$id']")?.parentElement
}
```

And then in the views themselves, I added an event listener after first render:

```kotlin
firstRender += {
    htmlElement?.addEventListener("contextmenu", {
        it.preventDefault()
    })
}
```

**Reply by pusolito** [reply]
OK, glad you have a work-around for now. I'll take a look at this and see if there's something missing to make this work on the pointer event--which would be the most portable (with non-browser platforms)--or some solution that incorporates `oncontextmenue`.

Filed #52.

**Reply by pusolito** [reply, marked as answer]
This will be properly supported in 0.9.2. The menu will be suppressed if you consume the pointer pressed event for a view.

```
view.pointerChanged += {
    // ...
    if (Button2 in it.buttons) {
        // This will prevent the context menu in 0.9.2
        it.consume()
    }
}
```

**Reply by pusolito** [reply]
@amarcolini can you give the [0.9.2 snapshot](https://kotlinlang.slack.com/archives/C01CJM07MGV/p1684610892836579) build a try and see if it solves the issue?

**Reply by amarcolini** [reply]
@pusolito do I have to join the Kotlin Slack to access that build?

**Reply by pusolito** [reply]
No. You can try it out by pointing to maven { url = uri("https://oss.sonatype.org/content/repositories/snapshots") }

**Reply by pusolito** [reply]
There is also a new Menu control that might actually fit your use case if you plan on doing a multi-level menu.

The new Menu control is a View that contains a vertical list of interactive items. These items notify whenever the user interacts with them using a pointer or the keyboard. This control support a couple types of action items and one that shows a new "sub" menu when the user interacts with it. You specify the contents of a `Menu` using the `MenuFactory`, which provides a declarative DSL for defining the structure and behavior of Menus.

```kotlin
val menu = menus(close = { /* close menu */ }) {
    menu("Sub menu") {
        prompt   ("Prompt"      ) { /* handle */ }
        separator(              )
        action   ("Do something") { /* handle */ }
        menu     ("Sub-sub menu") {
            action   ("Some action"   ) { /* handle */ }
            separator(                )
            action   ("Another action") { /* handle */ }
        }
    }
    action("Execute action") { /* handle */ }
}
```

**Reply by amarcolini** [reply]
I've tried 0.9.2-SNAPSHOT, but running it gives me `Uncaught ReferenceError: Application is not defined` in the browser. The line it occurs on in the compiled js is `setMetadataFor(GUIApp, 'GUIApp', classMeta, undefined, [Application], undefined, undefined, []);`, where GUIApp is the name of my application class.

**Reply by pusolito** [reply]
Strange. Did you do a clean build? Also, are you using the JS IR compiler or Legacy?

**Reply by amarcolini** [reply]
I've been using the JS IR compiler. A clean build fixed the problem :facepalm:. Consuming the pressed event using 0.9.2-SNAPSHOT does indeed prevent context menus. Thank you for the fix! I'll be sure to look at the Menu control.

**Reply by pusolito** [reply]
Just curious; is this work your doing to update Joos? Or a different project?

**Reply by amarcolini** [reply]
I'm currently overhauling the GUI for Joos, among other improvements. I just haven't made any commits in forever.
